nomic/embed.py

```diff
@@ -220,32 +232,31 @@ def _text_embed4all(
     task_type: str,
     dimensionality: int | None,
     long_text_mode: str,
+    dynamic_mode: bool,
     **kwargs: Any,
 ) -> dict[str, Any]:
-    global _embed4all, _embed4all_kwargs, _embed4all_has_init_gpu
+    global _embed4all, _embed4all_kwargs
 
     try:
         g4a_model = _EMBED4ALL_MODELS[model]
     except KeyError:
         raise ValueError(f"Unsupported model for local embeddings: {model!r}") from None
 
-    if any(text == "" for text in texts):
-        raise NotImplementedError("Embedding an empty text is not implemented")
-
     if not texts:
         # special-case this since Embed4All doesn't allow it
-        return {"embeddings": [], "usage": {}, "model": model}
+        return {"embeddings": [], "usage": {}, "model": model, "inference_mode": "local"}
 
     if _embed4all is None or _embed4all.gpt4all.config["filename"] != g4a_model or _embed4all_kwargs != kwargs:
-        using_gpu = kwargs.get("device") not in (None, "cpu")
-        if using_gpu and _embed4all_has_init_gpu:
-            raise NotImplementedError("GPU context can only be initialized once")
-
+        if _embed4all is not None:
+            _embed4all.close()
         _embed4all = Embed4All(g4a_model, **kwargs)
         _embed4all_kwargs = kwargs
 
-        if using_gpu:
-            _embed4all_has_init_gpu = True
+    def cancel_cb(batch_sizes: list[int], backend: str) -> bool:
+        # TODO(jared): make this more accurate
+        n_tokens = sum(batch_sizes)
+        limits = {"cpu": 16, "kompute": 32, "metal": 1024}
```

Comment:
how did you settle on these numbers?

Reply:
I plotted the number of tokens in one batch against the time taken to generate the embeddings, and got this:

<img width="940" alt="Screenshot 2024-04-12 at 14 45 27" src="https://github.com/nomic-ai/nomic/assets/14168726/e135c1da-b6f5-42e7-a875-d98548b9b8c7">

So I chose thresholds at the 2^4, 2^5, and 2^10 input sizes.